skyverge/wc-plugin-framework
# Remove Framework table

# Summary

This PR removes calls to methods that are no longer necessary now that framework tokens are included in WooCommerce's table of payment methods.

### Story: [CH 30010](https://app.clubhouse.io/skyverge/story/30010/remove-framework-table)
### Release: #362

## Details

- I didn't remove leftover CSS and JS because we are still making small modifications here and there. 
- Tokens are no longer loaded on init with `load_tokens()`
- The framework's table is no longer rendered (via `render()`) at the bottom of the table
- `handle_payment_method_actions()` is no longer used because actions are now handled in core
- `render()` and `handle_payment_method_actions()` are now deprecated
- The following actions are now mapped to similar core actions for backwards compatibility
    - wc_{plugin_id}_before_my_payment_method_table
    - wc_{plugin_id}_after_my_payment_method_table
    - wc_payment_gateway_{gateway_id}_payment_method_deleted
- Restored support for custom payment methods actions via the  `wc_{plugin_id}_my_payment_methods_table_method_actions` and `wc_{plugin_id}_my_payment_methods_action_{action}` actions
- Deprecated several methods that render HTML for the framework table of payment methods

## QA

### Setup

- Configure NETBilling to use `dev-ch30010-remove-framework-table` version of the plugin framework (this branch)
- Have at least one saved payment method from the framework

### Steps

#### Check the table is displayed correclty and JavaScript behavior is working

1. Navigate to My Payment Methods
    - [x] The Title, Details, and Default columns show appropriate values for the payment method
1. Click the Edit button
    - [x] The "Title" column switches to a text input
    - [x] The "Edit" button switches to the "Cancel" link
1. Enter a nickname for the payment method
1. Click Save
    - [x] The "Title" column switches back to plain text and displays the new nickname
    - [x] The "Cancel" link switches back to the "Edit" button

#### Test mapped actions

Add the following snippet to your site:

```php
add_action( 'wc_netbilling_before_my_payment_method_table', function() {

    echo '<div class="woocommerce-message">BEFORE</div>';
} );


add_action( 'wc_netbilling_after_my_payment_method_table', function() {

    echo '<div class="woocommerce-message">AFTER</div>';
} );


add_action( 'wc_payment_gateway_netbilling_payment_method_deleted', function() {

    wc_add_notice( 'wc_payment_gateway_netbilling_payment_method_deleted action was fired!' );
} );
```

1. Navigate to My Payments Methods
    - [x] A notice that reads "BEFORE" is visible before the table
    - [x] A notice that reads "AFTER" is visible after the table
1. Delete a framework saved payment method
    - [x] A notice with "wc_payment_gateway_netbilling_payment_method_deleted action was fired!" is shown when the page reloads

### Test custom payment methods actions

Add the following snippet to your site:

```php
add_action( 'wc_netbilling_my_payment_methods_table_method_actions', function( $actions, $token, $instance ) {

    $plugin_slug = $instance->get_plugin()->get_id_dasherized();

    $url = add_query_arg( [
        "wc-{$plugin_slug}-token"  => $token->get_id(),
        "wc-{$plugin_slug}-action" => 'custom',
    ] );

    $actions['custom'] = [
        'name' => 'Custom',
        'url'  => wp_nonce_url( $url, "wc-{$plugin_slug}-token-action" ),
    ];

    return $actions;
}, 10, 3 );

add_action( 'wc_netbilling_my_payment_methods_action_custom', function( $instance ) {

    wc_add_notice( 'wc_netbilling_my_parent_methods_action_custom action was fired!' );
} );
```

1. Navigate to My Payments Methods
    - [x] NETBilling payment methods have a Custom action button
1. Click the Custom action button
    - [x] A notice with "wc_netbilling_my_parent_methods_action_custom action was fired!" is shown when the page reloads
